torch/_inductor/fx_passes/mkldnn_fusion.py

```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+            )
+        packed_conv_inputs = (
+            (input, packed_weight_node, bias) + tuple(constant_args) + ("none", [], "")
+        )
+        packed_conv_node = gm.graph.create_node(
+            "call_function", packed_conv_op, tuple(packed_conv_inputs)
+        )
+        conv_node.replace_all_uses_with(packed_conv_node)
+        packed_conv_node.meta.update(conv_node.meta)
+        gm.graph.erase_node(conv_node)
+
+
+def _insert_packed_linear(gm, linear_node):
+    weight_idx = 2 if linear_node.target == aten.addmm.default else 1
+    input = linear_node.args[weight_idx - 1]
+    batch_size = input.meta.get("val").shape[0]
+    permute_node = linear_node.args[weight_idx]
+    linear_weight = permute_node.args[0]
+    linear_bias = linear_node.args[0] if weight_idx == 2 else None
+    with gm.graph.inserting_before(linear_node):
+        mkldnn_tensor_node = gm.graph.create_node(
+            "call_method", "to_mkldnn", (linear_weight,)
+        )
+        is_bf16_weight = linear_weight.meta.get("val").dtype == torch.bfloat16
+        if not is_bf16_weight and free_symbols(batch_size):
+            return
+        if free_symbols(batch_size):
+            # TODO: support symbolic input size for bfloat16 path
+            return
+        else:
+            packed_weight_inputs = (mkldnn_tensor_node, batch_size)
+            packed_weight_op = (
+                torch.ops.mkldnn._reorder_linear_weight
+                if is_bf16_weight
+                else torch.ops.mkl._mkl_reorder_linear_weight
+            )
+            packed_weight_node = gm.graph.create_node(
+                "call_function", packed_weight_op, args=packed_weight_inputs
+            )
+        packed_linear_inputs = (input, packed_weight_node)
+        if is_bf16_weight:
+            packed_linear_inputs += (linear_bias, "none", [], "")
+            packed_linear_op = torch.ops.mkldnn._linear_pointwise.default
+        else:
+            packed_linear_inputs += (linear_weight, linear_bias, batch_size)
+            packed_linear_op = torch.ops.mkl._mkl_linear
+        packed_linear_node = gm.graph.create_node(
+            "call_function", packed_linear_op, packed_linear_inputs
+        )
+        linear_node.replace_all_uses_with(packed_linear_node)
+        packed_linear_node.meta.update(linear_node.meta)
+        gm.graph.erase_node(linear_node)
+        if len(permute_node.users) <= 1:
+            gm.graph.erase_node(permute_node)
+
+
+def _eliminate_duplicate_packed_nodes(gm):
+    """
+    Combine packed weight nodes with the same inputs to reduce memory usage.
+    for example:
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = nn.Linear(32, 32, bias=True)
+
+        def forward(self, x):
+            return self.linear(self.linear(x))
+
+    the above's packed weight nodes are duplicate if two linear calls have same input size.
```

Comment:
Can this duplication be avoided in the `graph.py` when the constant is added to the graph with `add_tensor_constant`?

Reply:
We can't do it at ```add_tensor_constant``` which MKLDNN tensor doesn't support comparison.

Reply:
It seems the dedup logic is not necessarily specific to mkldnn weight packing patterns but can be made general, i.e., sharing the common prefix of users, e.g.,
```
w -> op1 -> op2 -> op3
  -> op1 -> op2 -> op4
```
Can be simplified to
```
w -> op1 -> op2 -> op3
                -> op4
```